src/runtime/gc_list.go

```diff
@@ -0,0 +1,460 @@
+			interrupt.Restore(i)
+			if gcDebug {
+				println("scanning tasks queued during GC")
+			}
+			goto runqueueScan
+		}
+		runqueue = markedTaskQueue
+		interrupt.Restore(i)
+	}
+
+	if gcDebug {
+		println("GC mark phase complete")
+		println("remaining allocations:")
+		remainingAllocs.dump()
+	}
+
+	return remainingAllocs
+}
+
+// sort the allocation list in ascending address order.
+func (list *allocList) sort() {
+	// Sort the list by repeatedly moving the lowest-address element to the front.
+	// This is selection sort (https://en.wikipedia.org/wiki/Selection_sort).
+	for dst := &list.first; *dst != nil; dst = &((*dst).next) {
+		// Find the next lowest-address allocation.
+		src := dst
+		min := *src
+		for ref, alloc := src, min; alloc != nil; ref, alloc = &alloc.next, alloc.next {
+			if uintptr(unsafe.Pointer(alloc)) < uintptr(unsafe.Pointer(min)) {
+				src, min = ref, alloc
+			}
+		}
+
+		// Remove the min from the list.
+		*src = min.next
+
+		// Insert the min at dst.
+		*dst, min.next = min, *dst
+	}
+
+	if gcDebug {
+		println("sorted allocations:")
+		list.dump()
+	}
+
+	if gcAsserts && list.first != nil {
+		prev := list.first
+		for alloc := list.first.next; alloc != nil; alloc, prev = alloc.next, alloc {
+			if uintptr(unsafe.Pointer(alloc)) <= uintptr(unsafe.Pointer(prev)) {
+				runtimePanic("alloc list not sorted")
+			}
+		}
+	}
+
+	if gcAsserts {
+		var prevEnd uintptr
+		for alloc := list.first; alloc != nil; alloc, prevEnd = alloc.next, uintptr(unsafe.Pointer(alloc))+unsafe.Sizeof(allocHeader{})+alloc.len {
+			start := uintptr(unsafe.Pointer(alloc))
+			if start < prevEnd {
+				runtimePanic("overlapping allocations")
+			}
+		}
+	}
+}
+
+// scan the allocation for pointers to other allocations.
+func (alloc *allocHeader) scan() {
+	start := uintptr(unsafe.Pointer(alloc)) + unsafe.Sizeof(allocHeader{})
+	end := start + alloc.len
+
+	if gcDebug {
+		println("scanning allocation at", start, "of", alloc.len, "bytes")
+	}
+
+	markRoots(start, end)
+}
+
+// markRoots reads all pointers from start to end (exclusive) and if they look
+// like a heap pointer and are unmarked, marks them and scans that object as
+// well (recursively). The start and end parameters must be valid pointers and
+// must be aligned.
+func markRoots(start, end uintptr) {
+	if gcDebug {
+		println("mark from", start, "to", end, int(end-start))
+	}
+	if gcAsserts {
+		if start >= end {
+			runtimePanic("gc: unexpected range to mark")
+		}
+	}
+
+	for addr := start; addr+unsafe.Sizeof(unsafe.Pointer(nil)) <= end; addr += unsafe.Alignof(addr) {
+		root := *(*uintptr)(unsafe.Pointer(addr))
+		markRoot(addr, root)
+	}
+}
+
+// mark a GC root at the address addr.
+func markRoot(addr, root uintptr) {
+	if !looksLikePointer(root) {
+		// This is not a pointer.
+		// Skip searching for it.
+		return
+	}
+
+	// Search the list of unmarked allocations for this root address.
+	alloc := activeAllocs.search(root)
```

Comment:
This also looks like a quadratic algorithm here: for every possible root, all allocations are scanned.

Reply:
Unfortunately yes.